### ISSUE TITLE
How to rebuild snapshot based on Events?

### ISSUE THREAD

**Issue by tomaszferens** [rrweb-io/rrweb#1153]
I want to rebuild a snapshot/website based on the recorded events:

```
let events = [];

const stop = rrweb.record({
  emit(event) {
    events.push(event);
  },
});

setTimeout(() => {
  stop();
  // ?? Reconstruct the entire recorded HTML based on the events array
  // Reconstruct at the point of stopping the recording, so to the last event
}, 5000)
```

Does rrweb expose some util/helper that does this? I would like to get the entire HTML string

**Comment by YunFeng0817** [comment]
I guess you need this package: https://github.com/rrweb-io/rrweb/tree/master/packages/rrweb-snapshot

**Comment by tomaszferens** [comment]
@YunFeng0817 does it support events array? I would prefer to do it using events array by applying a diff so that it is performant - imagine taking full snapshot each second. Is that possible? Maybe I would need to dig into rrweb-player source code because I believe that you do the same technique there.

**Comment by YunFeng0817** [comment]
https://github.com/rrweb-io/rrweb/blob/master/guide.md
Would you like this option in rrweb? 
<img width="1174" alt="image" src="https://user-images.githubusercontent.com/27533910/221107787-4ee6a56f-4c28-44ac-a9a5-f2c83736c6c7.png">
Sorry, I'm not quite sure about your scenario.

**Comment by tomaszferens** [comment]
Ohh I didn't see that option, nice! Yes, I would need something like that. Is it possible to manually decide when to checkout? Instead of doing it by "N ms" or "N events"?

My scenario is that I would like to record + take a full snapshot only after user clicked/interacted with a page in any way.

**Comment by YunFeng0817** [comment]
Yes, I believe this function will suit your needs.
https://github.com/rrweb-io/rrweb/blob/master/packages/rrweb/src/record/index.ts#L636

**Comment by tomaszferens** [comment]
Thank you so much. Will look into this! I assume this will generate the `FullSnapshot` event type, and now to rebuild it I need to use `rebuild` function from `rrweb-snapshot`?

**Comment by YunFeng0817** [comment]
Yes, it will generate a Fullsnapshot event. 
I start to understand your needs now. 
Basically, a full snapshot is like a photo and the recorded data of rrweb is like a video. If you only need photos rather than videos, then the rrweb-snapshot can do the job. The package doesn't care about the storage of fullsnapshot shots.
To record the snapshot, you can use the function snapshot https://github.com/rrweb-io/rrweb/blob/master/packages/rrweb-snapshot/README.md#snapshot. And to rebuild the page, you may use the function https://github.com/rrweb-io/rrweb/blob/master/packages/rrweb-snapshot/README.md#rebuild.

**Comment by tomaszferens** [comment]
I see, I will take a look at `rrweb-snapshot`. The reason I wanted to do it via `rrweb.record()` and `events` is because I thought that it's gonna be more performant than taking fullsnapshot + rebuilding it. Usually when you interact with a page not much has changed (let's say you clicked on a button to open dropdown menu). So I wanted to take the advantage of the `events` and somehow apply a diff to reconstruct new snapshot

**Comment by YunFeng0817** [comment]
I think you may record a full video of your session and only keep the interaction time somewhere. In the playback,  you can use play(timeOffset) function of Replayer that can directly jump to that time without playing intermediate frames. Just like the seek feature in video players.

**Comment by tomaszferens** [comment]
Yes, that sounds very promising. I will definitely look into this approach. Is it possible to grab the HTML string out of the "frame" (`play(timeOffset)`) somehow? I will also look into Replayer to better understand how do you reconstruct website based on the events.

**Comment by YunFeng0817** [comment]
I think it's possible to grab HTML by innerHTML
```
replayer.iframe.contentDocument.documentElement.innerHTML
```
<img width="448" alt="image" src="https://user-images.githubusercontent.com/27533910/221125482-73ffe4fc-b2f4-4502-86c8-c41dd9fd93f5.png">

**Comment by Misterion777** [comment]
> replayer.iframe.contentDocument.documentElement.innerHTML

Sometimes, `replayer.iframe` object is undefined, what we should do in this case?
